ruby_version_file_parser_test.go

```diff
@@ -0,0 +1,101 @@
+package bundleinstall_test
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+
+	bundleinstall "github.com/paketo-buildpacks/bundle-install"
+	"github.com/sclevine/spec"
+
+	. "github.com/onsi/gomega"
+)
+
+const RUBY_VERSION_FILE_TEMPLATE = "%s\n"
+
+func testRubyVersionFileParser(t *testing.T, context spec.G, it spec.S) {
+	var (
+		Expect = NewWithT(t).Expect
+
+		path   string
+		parser bundleinstall.RubyVersionFileParser
+	)
+
+	it.Before(func() {
+		file, err := os.CreateTemp("", ".ruby-version")
+		Expect(err).NotTo(HaveOccurred())
+		defer file.Close()
+
+		path = file.Name()
+
+		parser = bundleinstall.NewRubyVersionFileParser()
+	})
+
+	it.After(func() {
+		Expect(os.RemoveAll(path)).To(Succeed())
+	})
+
+	context("ParseVersion", func() {
+		context("when given different types of versions", func() {
+			var versions []string
+			it.Before(func() {
+				versions = []string{
+					"2",
+					"2.6",
+					"2.6.0",
+					"2.7.0",
+					"2.10.0",
+					"2.10.10",
+					"10.0.0",
+				}
+			})
+			it("parses the versions correctly", func() {
+				for _, v := range versions {
+					Expect(os.WriteFile(path, []byte(fmt.Sprintf(RUBY_VERSION_FILE_TEMPLATE, v)), 0644)).To(Succeed())
```

Comment:
should this bit go in the Before too?

Reply:
Ah I misunderstood this a bit, sorry. As the file needs to be written for every version I have moved the `it.Before` inside the loop as well and moved the versions var into the `vars` block. At least to me this looks cleaner now but I'm not super experienced with gomega tests, so this might not be the most idiomatic way :)

Reply:
This BDD style of testing doesn't lend itself very cleanly to the iterative style that you've written. Ideally BDD wants for one assertion per `it()`. We often group a set of assertions together because they're logically related. For example, if a method returns `(x, y, err)`, we often want to test that `x` and `y` have the correct values and that `err` is `nil`). We'd also assert for any relevant side effects too.

However, different inputs (`versions` in this situation) are different test cases, and so the most idiomatic way to write these tests would be one `it` per test case.

One benefit of breaking out the test cases this way is that you get to write a different description for each case, which helps with debugging failing test cases and more generally understanding the behavior of the application under test.

For example, in this case I'd expect to see something like the following:

```
context("when the version is just a single Major value", func() {
  it.Before(func() {
    Expect(os.WriteFile(path, "2\n", 0644)).To(Succeed())
  })

  it("parses correctly", func() {
    version, err := parser.ParseVersion(path)
	Expect(err).NotTo(HaveOccurred())
	Expect(version).To(Equal("2"))
  })
})

context("when the version is a Major.Minor value", func() {
  it.Before(func() {
    Expect(os.WriteFile(path, "2.6\n", 0644)).To(Succeed())
  })

  it("parses correctly", func() {
    version, err := parser.ParseVersion(path)
	Expect(err).NotTo(HaveOccurred())
	Expect(version).To(Equal("2.6"))
  })
})
```

Another benefit of this is it forces you to only consider the necessary surface area for testing. Specifically, I don't think we need to test `2.6.0` and `2.7.0` because they're structurally identical - there's nothing special about `2.7.0` that doesn't also apply to `2.6.0`.

I acknowledge this is really verbose, but then so is golang :)

Reply:
Thanks, that makes sense. I adjusted the test. Just want to mention that I initially took some inspiration from the test in `gemfile_parser_test.go` 😄